### PR TITLE
Added command store {NAME} {REPLACE} to save the player queue to an existing playlist

### PR DESCRIPTION
**Pull request by geneticdrift**
by either replacing or appending to the existing list.

### CODE REVIEW

**Comment by MaxKellermann** [comment]
I agree that both features added are useful, but I find the protocol design rather clumsy. We have "save" which does one thing, and now we have "store" (a word which has roughly the same subjective meaning as "save) which does two slightly different things, chosen by one obscure parameter.
I know the MPD protocol has accumulated quite a bit of obscurity and inconsistency across the last 19 years, but I think new commands should try to be better.
Maybe the mode should just be the second parametr to the existing "save" command? Just a rough draft:
- `save FOO` is the same as `save FOO create` (for backwards compatibility)
- `save FOO replace`
- `save FOO append` (does `append` allow creating a new file if it does not yet exist? Do we need both modes? Create-or-append and append-only-and-fail-if-not-exists?)

**Comment by geneticdrift** [comment]
For `append` and `replace` I think the expectation is that the playlist `FOO` already exists. So following your draft I've changed the implementation to use the `save` command with an optional `MODE` argument, where `create` is the default, and the other optional values are `replace` and `append`.

**Comment by jcorporation** [comment]
Great new feature. I had in the past already feature requests to implement this on client side.
As soon this feature is merged I will implement support for it in libmpdclient and in myMPD!

File: doc/protocol.rst [hunk 1]
```diff
@@ -1006,21 +1006,20 @@ remote playlists (absolute URI with a supported scheme).
...
-:command:`save {NAME}`
+:command:`save {NAME} [MODE]`
     Saves the queue to
     `NAME.m3u` in the playlist directory.
 
     If a playlsit with name ``NAME`` already exists, the command fails with "Playlist already exists".
 
-.. _command_store:
+    ``MODE``
+        Optional argument [#since_0_xx_yy]_. One of `create`, `append`, or `replace`.
```
Comment (MaxKellermann) [inline doc/protocol.rst:1016]: since_0_xx_yy?

File: doc/protocol.rst [hunk 2]
```diff
@@ -1006,21 +1006,20 @@ remote playlists (absolute URI with a supported scheme).
...
 
-:command:`store {NAME} {REPLACE}`
-    Saves the queue to an existing
-    `NAME.m3u` in the playlist directory.
-
-    If a playlsit with name ``NAME`` doesn\'t already exist, the command fails with "No such playlist".
+        `create`
+            The default. If a playlist with name ``NAME`` already exists, the command fails with "Playlist already exists".
```
Comment (MaxKellermann) [inline doc/protocol.rst:1019]: Actual error messages should not be documented, or else we can't ever rephrase them, because then clients will rely on the exact wording.

File: src/PlaylistSave.hxx [hunk 3]
```diff
@@ -31,32 +31,38 @@ playlist_print_song(BufferedOutputStream &os, const DetachedSong &song);
...
-/**
- * Saves a queue object into a stored playlist file.
- */
-void
-spl_save_queue(const char *name_utf8, const Queue &queue);
+struct PlaylistSaveMode {
+private:
+	bool create = false;
+	bool replace = false;
+	bool append = false;
```
Comment (MaxKellermann) [inline src/PlaylistSave.hxx:38]: This is a rather complicated and bloated way to store this - why not an enum?

File: src/PlaylistSave.hxx [hunk 4]
```diff
@@ -31,32 +31,38 @@ playlist_print_song(BufferedOutputStream &os, const DetachedSong &song);
...
- * Saves a playlist object into a stored playlist file.
- */
-void
-spl_save_playlist(const char *name_utf8, const playlist &playlist);
+	const char *argValidationError = "Expected one of 'create', 'append', 'replace'";
+
+public:
+	[[nodiscard]] static PlaylistSaveMode parseArg(const char* arg);
+
+	[[nodiscard]] inline bool is_valid() const noexcept {
```
Comment (MaxKellermann) [inline src/PlaylistSave.hxx:45]: Redundant `inline`, and please use CamelCase for method/class names. (Very old code often still uses lower-case with underscores.)

File: src/PlaylistSave.hxx [hunk 5]
```diff
@@ -31,32 +31,38 @@ playlist_print_song(BufferedOutputStream &os, const DetachedSong &song);
...
+		return create || (append != replace);
+	}
+
+	[[nodiscard]] inline bool is_create_only() const noexcept {
+		return create && !(append || replace);
```
Comment (MaxKellermann) [inline src/PlaylistSave.hxx:50]: This... looks complicated.

File: src/PlaylistSave.hxx [hunk 6]
```diff
@@ -31,32 +31,38 @@ playlist_print_song(BufferedOutputStream &os, const DetachedSong &song);
...
+};
 
-/**
- * Store a queue object into a stored playlist file.
- *
- * @arg replace true to relace the stored playlist contents, false to append to the list.
- */
 void
-spl_store_queue(const char *name_utf8, bool replace, const Queue &queue);
+spl_save_queue(const char *name_utf8, const PlaylistSaveMode& save_mode, const Queue &queue);
```
Comment (MaxKellermann) [inline src/PlaylistSave.hxx:63]: Pass this by value, not by reference - hey, even if you keep this as struct with 3 bools, it's just 3 bytes (1 bytes for an `enum : uint8_t`) - the reference will take 8 bytes for the pointer, plus the size of the struct!

File: src/PlaylistSave.cxx [hunk 7]
```diff
@@ -83,17 +85,44 @@ playlist_print_uri(BufferedOutputStream &os, const char *uri)
...
+		if (StringIsEqual(arg, "create"))
+			mode.create = true;
+		else if (StringIsEqual(arg, "append"))
+			mode.append = true;
+		else if (StringIsEqual(arg, "replace"))
+			mode.replace = true;
+	}
+
+	if (!mode.is_valid())
+		throw MakeArgError(mode.argValidationError, arg == nullptr ? "" : arg);
```
Comment (MaxKellermann) [inline src/PlaylistSave.cxx:103]: This whole protocol-specific method living in a non-protocol-specific library is kind of a layering violating - I suggest we leave it that way for now, but please use `std::invalid_argument`; if caught by a command handler, it will be translated to ACK_ERROR_ARG, so that's fine:
https://github.com/MusicPlayerDaemon/MPD/blob/512cd7b0dea47883e54245122171a0178678e178/src/command/CommandError.cxx#L99-L100

File: src/PlaylistSave.hxx [hunk 8]
```diff
@@ -31,32 +31,38 @@ playlist_print_song(BufferedOutputStream &os, const DetachedSong &song);
...
+
+	[[nodiscard]] inline bool is_valid() const noexcept {
+		return create || (append != replace);
+	}
+
+	[[nodiscard]] inline bool is_create_only() const noexcept {
+		return create && !(append || replace);
+	}
+
+	[[nodiscard]] inline bool is_existing_only() const noexcept {
```
Comment (MaxKellermann) [inline src/PlaylistSave.hxx:53]: More redundant `inline` - `inline` is implied for method bodies inside `struct{}`!

**Review by MaxKellermann — CHANGES_REQUESTED** [review]
A few nitpicks.
The second commit is just a fixup for the first one; please fold them into one.

File: src/PlaylistSave.cxx [hunk 9]
```diff
@@ -84,20 +84,29 @@ playlist_print_uri(BufferedOutputStream &os, const char *uri)
...
+	}
+	else if (!FileExists(path_fs)) {
+		throw PlaylistError(PlaylistResult::NO_SUCH_LIST, "No such playlist");
+	}
 
-	FileOutputStream fos(path_fs);
+	FileOutputStream fos(path_fs,
+						 save_mode == PlaylistSaveMode::APPEND
+										 ? FileOutputStream::Mode::APPEND_EXISTING
+										 : FileOutputStream::Mode::CREATE);
```
Comment (MaxKellermann) [inline src/PlaylistSave.cxx:104]: Bad indent

File: src/PlaylistSave.cxx [hunk 10]
```diff
@@ -84,20 +84,29 @@ playlist_print_uri(BufferedOutputStream &os, const char *uri)
...
 	BufferedOutputStream bos(fos);
 
-	for (unsigned i = 0; i < queue.GetLength(); i++)
+	for (unsigned i = 0; i < queue.GetLength(); i++) {
 		playlist_print_song(bos, queue.Get(i));
+	}
```
Comment (MaxKellermann) [inline src/PlaylistSave.cxx:109]: Don't change existing code style (in an unrelated commit); this adds unnecessary noise to the diff.

File: src/PlaylistSave.cxx [hunk 11]
```diff
@@ -106,7 +115,7 @@ spl_save_queue(const char *name_utf8, const Queue &queue)
 }
 
 void
-spl_save_playlist(const char *name_utf8, const playlist &playlist)
+spl_save_playlist(const char *name_utf8, PlaylistSaveMode save_mode, const playlist &playlist)
 {
-	spl_save_queue(name_utf8, playlist.queue);
+    spl_save_queue(name_utf8, save_mode, playlist.queue);
```
Comment (MaxKellermann) [inline src/PlaylistSave.cxx:120]: Bad indent

File: src/PlaylistSave.hxx [hunk 12]
```diff
@@ -31,16 +31,16 @@ playlist_print_song(BufferedOutputStream &os, const DetachedSong &song);
 void
 playlist_print_uri(BufferedOutputStream &os, const char *uri);
 
-/**
- * Saves a queue object into a stored playlist file.
- */
```
Comment (MaxKellermann) [inline src/PlaylistSave.hxx:36]: Why delete API documentation?

**Review by MaxKellermann — CHANGES_REQUESTED** [review]
Some coding style nitpicks; I think after that I'll merge it.
Before you re-push a PR branch, please always rebase on the official master branch.
